compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPU2DScalableTo1DScalable.cpp

```diff
@@ -0,0 +1,175 @@
+///     for n in range(0, 4 x vscale) step 4 {
+///        // Drop a scalable dim:
+///        linalg.generic {lowering_config = [4, [4]]
+///     }
+///   }
+/// }
+/// ```
+///
+/// This can now be vectorized and lowered successfully, which produces a
+/// dispatch that mixes SME and SVE.
+class LLVMCPU2DScalableTo1DScalablePass
+    : public LLVMCPU2DScalableTo1DScalableBase<
+          LLVMCPU2DScalableTo1DScalablePass> {
+public:
+  using LLVMCPU2DScalableTo1DScalableBase::LLVMCPU2DScalableTo1DScalableBase;
+
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry
+        .insert<arith::ArithDialect, linalg::LinalgDialect, scf::SCFDialect>();
+  }
+
+  void runOnOperation() override;
+};
+
+static bool opKnownToSupport2DScalableVectorizationWithArmSME(Operation *op) {
+  return isa<linalg::MatmulOp, linalg::MatmulTransposeAOp, linalg::FillOp>(op);
+}
+
+// Note: It would be easy to parameterize this rewrite to convert N-D scalable
+// operations to M-D scalable ones (where M < N). However this is currently not
+// needed.
+static LogicalResult
+dropScalabilityFromUnsupportedOperations(mlir::FunctionOpInterface funcOp,
+                                         bool assumeArmSME = false) {
+  // Note: Which operations should have scalability dropped is specific to
+  // ArmSME. The rest of this rewrite could be generic (though currently
+  // there's no other targets that support > 1D scalability).
+  auto targetAttr = IREE::HAL::ExecutableTargetAttr::lookup(funcOp);
+  bool isArmSME = assumeArmSME || hasSMEFeature(targetAttr);
+
+  if (!isArmSME)
+    return success();
+
+  SmallVector<TilingInterface> computeOps;
+  funcOp.walk([&](TilingInterface op) {
+    if (!opKnownToSupport2DScalableVectorizationWithArmSME(op))
+      computeOps.push_back(op);
+  });
+
+  for (TilingInterface tilingOp : computeOps) {
+    auto loweringConfigAttr = getLoweringConfig(tilingOp);
+    if (!loweringConfigAttr)
+      continue;
+
+    TilingConfig tilingConfig(loweringConfigAttr);
+    auto [vectorSizes, scalableFlags] = tilingConfig.getVectorTileSizes();
+    auto numScalableDims = llvm::count(scalableFlags, true);
+
+    if (numScalableDims <= 1)
+      continue;
+
+    SmallVector<int64_t> loopTileSizes;
+    SmallVector<bool> newScalableFlags;
+    for (auto [flag, size] : llvm::zip_equal(scalableFlags, vectorSizes)) {
+      if (flag && numScalableDims >= 2) {
+        --numScalableDims;
+        loopTileSizes.push_back(size);
+        newScalableFlags.push_back(false);
+      } else {
+        loopTileSizes.push_back(0);
+        newScalableFlags.push_back(flag);
+      }
+    }
+
+    IRRewriter rewriter(tilingOp->getContext());
+    rewriter.setInsertionPoint(tilingOp);
+
+    // 2. Re-tile the operation with some scalability dropped. This introduces
+    // loops for previously scalable vector/tile sizes.
+    scf::SCFTilingOptions options{};
+    setSCFTileSizes(options, tilingOp, loopTileSizes, {});
```

Comment:
nit: add function argument comment for better readability

```suggestion
    setSCFTileSizes(options, tilingOp, loopTileSizes, /*tileScalableFlags=*/{});
```